File: Homework/Manh/lesson-1/RangeSumQueryImmutable.java
class NumArray {

    int[] sums;
    
    public NumArray(int[] nums) {
        this.sums = new int[nums.length];
        sums[0] = nums[0];
        for(int i = 1; i < nums.length; i++) {
            sums[i] = nums[i] + sums[i-1];
        }
    }
    
    public int sumRange(int left, int right) {
        return sums[right] - (left == 0 ? 0 : sums[left-1]);
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(left,right);
 */

File: Homework/Khang_Nguyen/range-sum-query-immutable.js
/**
 * @param {number[]} nums
 */
 var NumArray = function(nums) {
    this.nums = nums
    const len = nums.length
    let tempSum = 0
    this.prefixSum = []
    for (let i = 0; i < len; i++)  {
        tempSum = tempSum + nums[i]
        this.prefixSum.push(tempSum)
    }
};

/** 
 * @param {number} left 
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function(left, right) {
   if (left === 0) return this.prefixSum[right]
   return this.prefixSum[right] - this.prefixSum[left - 1]
};

/** 
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */